### PR TITLE
nixos/activation: propagate system to nested configurations

### LINKED ISSUE #80806: nesting.clone is broken on non-x86_64
**Describe the bug**
The NixOS option `nesting.clone` is broken on non-x86_64 systems.
The problem is the `/run/current-system/fine-tune/child-1/bin/switch-to-configuration` script has shebang pointing to a x86_perl interpreter, regardless of the parent.

**To Reproduce**
1. `$ nix-build -A pkgsi686Linux.nixosTests.nesting`

**Expected behavior**
The script should have a shebang containing the right interpreter for the system.

**Additional context**
I think this has been broken since the beginning of NixOS.

Maintainer information:
```yaml
# a list of nixos modules affected by the problem
module:
- nixos/modules/system/activation/top-level.nix
```

### PR DESCRIPTION
**Pull request by thefloweringash**
###### Motivation for this change

Fix the nesting tests for non-x86_64 on hydra, and the other tests that depend on nesting like nginx.

I'm not confident that this composes correctly with `nixpkgs.localSystem` and `nixpkgs.crossSystem`, so reviews and tests appreciated.

Fixes #80806

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by grahamc** [comment]
Can you be more specific about what problem this is fixing?

**Comment by thefloweringash** [comment]
I somehow omitted the issue on first posting. I've added it now. It's aimed at #80806.

**Comment by grahamc** [comment]
This is amazing, I've never heard of `nesting`.

**Comment by worldofpeace** [comment]
> This is amazing, I've never heard of `nesting`.

Hehe, it's very useful to have an escape hatch configuration.

**Comment by worldofpeace** [comment]
@GrahamcOfBorg test nesting

**Comment by rnhmjoj** [comment]
I also tested i686-linux. I have no idea how this fixed it: anyone care to explain it?

**Comment by thefloweringash** [comment]
NixOS evaluation starts by calling `nixos/lib/eval-config.nix`. This takes a `system` parameter that defaults to `builtins.currentSystem`. The implementation of `nesting.clone` and `nesting.children` is based on calling this entry point again to evaluate an entirely separate NixOS, and then linking the result into the parent.

The problem was that the nested calls to the entry point did not set the `system` argument when calling `eval-config.nix`, which caused it to default to `builtins.currentSystem`. All nested configurations were built for the same system as the evaluating system. To fix this, it was sufficient to set the `system` to the same value as the top-level `eval-config.nix`, which is stored in `config.nixpkgs.system`.

There's an added complication in that the system values are encoded in the `pkgs` argument, which is derived from `config.nixpkgs.system`, `config.nixpkgs.localSystem`, `config.nixpkgs.crossSystem`, and `config.nixpkgs.pkgs` (for details, see [`nixos/modules/nixpkgs.nix`][]). Since there's no actual place where the value of `system` argument to `eval-config.nix` value is preserved, it's possible that some configurations will be surprising. In particular, `config.nixpkgs.system` will be inherited by all nested configurations, but the other parameters will not. This distinction likely only applies to `nesting.children`, which try not to inherit the parent configuration.

To prevent `config.nixpkgs.system` from being inherited unexpectedly, we could add a new internal option that stores the initial `system` passed to `eval-config.nix`, and use that for the nested calls.

[`nixos/modules/nixpkgs.nix`]: https://github.com/NixOS/nixpkgs/blob/806fac5d109cdc6653c33a18924dac31ac477a2b/nixos/modules/misc/nixpkgs.nix

**Comment by rnhmjoj** [comment]
Thank you, that was a very clear explanation.

> This distinction likely only applies to nesting.children, which try not to inherit the parent configuration.

Right, this change could cause some subtle problem, though I think the use of `nesting.children` is fairly uncommon.

>  the initial system passed to eval-config.nix, and use that for the nested calls.

That is because when `top-level.nix` is being evaluated the `system` argument stored in [config.nixpkgs.system](https://github.com/NixOS/nixpkgs/blob/806fac5d109cdc6653c33a18924dac31ac477a2b/nixos/lib/eval-config.nix#L43) could have been override by the user `configuration.nix`, right?

**Comment by thefloweringash** [comment]
> > the initial system passed to eval-config.nix, and use that for the nested calls.
> 
> That is because when `top-level.nix` is being evaluated the `system` argument stored in [config.nixpkgs.system](https://github.com/NixOS/nixpkgs/blob/806fac5d109cdc6653c33a18924dac31ac477a2b/nixos/lib/eval-config.nix#L43) could have been override by the user `configuration.nix`, right?

Yeah.

**Comment by rnhmjoj** [comment]
@thefloweringash can we go forward the internal option solution or do you have other plans?

**Comment by thefloweringash** [comment]
> @thefloweringash can we go forward the internal option solution or do you have other plans?

I've pushed an implementation of using the internal option

**Comment by rnhmjoj** [comment]
@GrahamcOfBorg test nesting nginx gitea

**Review by rnhmjoj — APPROVED** [review]
It looks good to me but I'm not an expert on this matter.

File: nixos/modules/misc/nixpkgs.nix [hunk 1]
```diff
@@ -216,6 +216,14 @@ in
         Ignored when <code>nixpkgs.pkgs</code> is set.
       '';
     };
+
+    initialSystem = mkOption {
+      type = types.str;
+      internal = true;
+      description = ''
+        Preserved value of `system` passed to `eval-config.nix`.
```
Comment (worldofpeace) [inline nixos/modules/misc/nixpkgs.nix]:
```suggestion
        Preserved value of <literal>system</literal> passed to <literal>eval-config.nix</literal>.
```

**Review by nbp — APPROVED** [review]
This sounds good to me.

At one moment I was wondering whether the `nested.clone` got the correct `nixpkgs.system` value, but as it is inheriting the same modules, it should be overwritten the same way.

Thanks!
(PS: I would not have noticed this issue without seeing the notification on IRC.)

**Comment by rnhmjoj** [comment]
@nbp Thank you for the review!

@thefloweringash could you squash the last two commit?

**Comment by rnhmjoj** [comment]
@GrahamcOfBorg test nesting nginx gitea

**Pull request merged**